### PR TITLE
Eventhubs Python preview2 merge to master

### PR DESCRIPTION
**Pull request by YijunXieMS**
Merge to master and get ready to release
detailed description to be added...

### CODE REVIEW

**Comment by adxsdk6** [comment]
Can one of the admins verify this patch?

File: sdk/eventhub/azure-eventhubs/azure/eventhub/_consumer_producer_mixin.py [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+
+from uamqp import errors, constants, compat
+from azure.eventhub.error import EventHubError, _handle_exception
+
+log = logging.getLogger(__name__)
+
+
+def _retry_decorator(to_be_wrapped_func):
+    def wrapped_func(self, *args, **kwargs):
+        timeout = kwargs.pop("timeout", None)
```
Comment (xiangyan99) [inline sdk/eventhub/azure-eventhubs/azure/eventhub/_consumer_producer_mixin.py]: Why not timeout = kwargs.pop("timeout", 100000)?

File: sdk/eventhub/azure-eventhubs/azure/eventhub/configuration.py [hunk 2]
```diff
@@ -8,7 +8,11 @@
 class _Configuration(object):
     def __init__(self, **kwargs):
         self.user_agent = kwargs.get("user_agent")
-        self.max_retries = kwargs.get("max_retries", 3)
+        self.retry_total = kwargs.pop('retry_total', 3)
+        self.max_retries = self.retry_total or kwargs.get("max_retries", 3)
```
Comment (xiangyan99) [inline sdk/eventhub/azure-eventhubs/azure/eventhub/configuration.py]: given retry_total has default value, kwargs.get("max_retries", 3) will never be used?

File: sdk/eventhub/azure-eventhubs/azure/eventhub/eventprocessor/checkpoint_manager.py [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+    """
+    def __init__(self, partition_id: str, eventhub_name: str, consumer_group_name: str, owner_id: str, partition_manager: PartitionManager):
+        self.partition_id = partition_id
+        self.eventhub_name = eventhub_name
+        self.consumer_group_name = consumer_group_name
+        self.owner_id = owner_id
+        self.partition_manager = partition_manager
+
+    async def update_checkpoint(self, offset, sequence_number=None):
+        """Users call this method in PartitionProcessor.process_events() to save checkpoints
```
Comment (xiangyan99) [inline sdk/eventhub/azure-eventhubs/azure/eventhub/eventprocessor/checkpoint_manager.py]: Do you need to put this module under aio folder?

File: sdk/eventhub/azure-eventhubs/azure/eventhub/eventprocessor/event_processor.py [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+
+        It provides the user a convenient way to receive events from multiple partitions and save checkpoints.
+        If multiple EventProcessors are running for an event hub, they will automatically balance loading. This feature
+        won't be available until preview 3.
+
+        :param eventhub_client: an instance of azure.eventhub.aio.EventClient object
+        :param consumer_group_name: the consumer group that is used to receive events
+        from the event hub that the eventhub_client is going to receive events from
+        :param partition_processor_factory: a callable (type or function) that is called to return a PartitionProcessor
+        TODO: elaborate an example
```
Comment (xiangyan99) [inline sdk/eventhub/azure-eventhubs/azure/eventhub/eventprocessor/event_processor.py]: TODO?

Comment (YijunXieMS) [reply sdk/eventhub/azure-eventhubs/azure/eventhub/eventprocessor/event_processor.py]: yes I'm still working on this

File: sdk/eventhub/azure-eventhubs/azure/eventhub/eventprocessor/checkpoint_manager.py (hunk 3 above)
Comment (YijunXieMS) [reply sdk/eventhub/azure-eventhubs/azure/eventhub/eventprocessor/checkpoint_manager.py]: eventprocessor has implied async for now. No sync eventprocessor

File: sdk/eventhub/azure-eventhubs/azure/eventhub/configuration.py (hunk 2 above)
Comment (YijunXieMS) [reply sdk/eventhub/azure-eventhubs/azure/eventhub/configuration.py]: good catch

File: sdk/eventhub/azure-eventhubs/azure/eventhub/_consumer_producer_mixin.py (hunk 1 above)
Comment (yunhaoling) [reply sdk/eventhub/azure-eventhubs/azure/eventhub/_consumer_producer_mixin.py]: good idea, will make the default value to 100000

File: sdk/eventhub/azure-eventhubs/azure/eventhub/eventprocessor/event_processor.py (hunk 4 above)
Comment (YijunXieMS) [reply sdk/eventhub/azure-eventhubs/azure/eventhub/eventprocessor/event_processor.py]: Updated this part and inserted an example

File: sdk/eventhub/azure-eventhubs/azure/eventhub/aio/client_async.py [hunk 5]
```diff
@@ -217,8 +222,12 @@ def create_consumer(
                 :caption: Add an async consumer to the client for a particular consumer group and partition.
 
         """
-        prefetch = self.config.prefetch if prefetch is None else prefetch
+        owner_level = kwargs.get("owner_level", None)
+        operation = kwargs.get("operation", None)
+        prefetch = kwargs.get("prefetch", None)
```
Comment (xiangyan99) [inline sdk/eventhub/azure-eventhubs/azure/eventhub/aio/client_async.py]: prefetch = kwargs.get("prefetch") or self.config.prefetch

File: sdk/eventhub/azure-eventhubs/azure/eventhub/client.py [hunk 6]
```diff
@@ -224,8 +228,11 @@ def create_consumer(
                 :caption: Add a consumer to the client for a particular consumer group and partition.
 
         """
-        prefetch = self.config.prefetch if prefetch is None else prefetch
+        owner_level = kwargs.get("owner_level", None)
+        operation = kwargs.get("operation", None)
+        prefetch = kwargs.get("prefetch", None)
```
Comment (xiangyan99) [inline sdk/eventhub/azure-eventhubs/azure/eventhub/client.py]: same

File: sdk/eventhub/azure-eventhubs/azure/eventhub/client_abstract.py [hunk 7]
```diff
@@ -266,6 +266,7 @@ def from_connection_string(cls, conn_str, event_hub_path=None, **kwargs):
                 :caption: Create an EventHubClient from a connection string.
 
         """
+        event_hub_path = kwargs.get("event_hub_path", None)
         is_iot_conn_str = conn_str.lstrip().lower().startswith("hostname")
         if not is_iot_conn_str:
             address, policy, key, entity = _parse_conn_str(conn_str)
```
Comment (xiangyan99) [inline sdk/eventhub/azure-eventhubs/azure/eventhub/client_abstract.py]: maybe pop event_hub_path here?

**Comment by yunhaoling** [comment]
@ramya-rao-a 

Hello Ramya, please review our readme and history file.

File: sdk/eventhub/azure-eventhubs/HISTORY.md [hunk 8]
```diff
@@ -1,5 +1,26 @@
...
+- Added new configuration parameters for exponential delay among each retry operation.
+    - `retry_total`: The total number of attempts to redo the failed operation.
+    - `backoff_factor`: The delay time factor.
+    - `backoff_max`: The maximum delay time in total.
+- Added support for context manager on `EventHubClient`.
+
+**Breaking changes**
+
+- Replaced `max_retries` configuration parameter of the EventHubClient with `retry_total`.
+- Introduced the initial concept of a new version of the `EventProcessor`, intended as a neutral framework for processing events across all partitions for a given Event Hub and in the context of a specific Consumer Group. This early preview is intended to allow consumers to test the new design using a single instance that does not persist checkpoints to any durable store.
```
Comment (ramya-rao-a) [inline sdk/eventhub/azure-eventhubs/HISTORY.md]: Why is this a "breaking change"? I would put this under "New features".

File: sdk/eventhub/azure-eventhubs/HISTORY.md [hunk 9]
```diff
@@ -1,5 +1,26 @@
 # Release History
 
+## 5.0.0b2 (2019-08-06)
+
+**New features**
+
+- Added new class `EventDataBatch` for publication of a batch of events with known size constraint.
+- Added new method `create_batch` to producer for creating EventDataBatch objects.
```
Comment (ramya-rao-a) [inline sdk/eventhub/azure-eventhubs/HISTORY.md]: In Javascript, we combined the above 2 points as they are related

> Added method `create_batch()` on the `EventHubProducer` to create an `EventDataBatch` that can then be used to add events until the maximum size is reached. This batch object can then be used in the `send()` method to send all the added events to Event Hubs.

Then in a sub bullet we have the below:

> This allows publishers to build batches without the possibility of encountering the error around the message size exceeding the supported limit when sending events. It also allows publishers with bandwidth concerns to control the size of each batch published.

File: sdk/eventhub/azure-eventhubs/HISTORY.md [hunk 10]
```diff
@@ -1,5 +1,26 @@
...
+- Added support for context manager on `EventHubClient`.
+
+**Breaking changes**
+
+- Replaced `max_retries` configuration parameter of the EventHubClient with `retry_total`.
+- Introduced the initial concept of a new version of the `EventProcessor`, intended as a neutral framework for processing events across all partitions for a given Event Hub and in the context of a specific Consumer Group. This early preview is intended to allow consumers to test the new design using a single instance that does not persist checkpoints to any durable store.
+    - `EventProcessor`: EventProcessor creates and runs consumers for all partitions of the eventhub.
+    - `PartitionManager`: PartitionManager defines the interface for getting/claiming ownerships of partitions and updating checkpoints.
+    - `PartitionProcessor`: PartitionProcessor defines the interface for processing events.
+    - `CheckpointManager`: CheckpointManager takes responsibility for updating checkpoints during events processing.
```
Comment (ramya-rao-a) [inline sdk/eventhub/azure-eventhubs/HISTORY.md]: This is a good idea to talk about the new classes being introduced.
@ShivangiReja, we should consider adding something similar

File: sdk/eventhub/azure-eventhubs/README.md [hunk 11]
```diff
@@ -206,6 +207,54 @@ finally:
     pass
 ```
 
+### Consume events from all partitions of an Event Hub
```
Comment (ramya-rao-a) [inline sdk/eventhub/azure-eventhubs/README.md]: As suggested in the README for Javascript and .Net, I would suggest to change this as below

```suggestion
### Consume events using an Event Processor
```

The idea is to catch the user's attention on how this is different from the other sample on consuming events.

File: sdk/eventhub/azure-eventhubs/HISTORY.md (hunk 8 above)
Comment (yunhaoling) [reply sdk/eventhub/azure-eventhubs/HISTORY.md]: thanks for the suggestion, doc is updated.

File: sdk/eventhub/azure-eventhubs/HISTORY.md (hunk 9 above)
Comment (yunhaoling) [reply sdk/eventhub/azure-eventhubs/HISTORY.md]: thanks for the suggestion, doc is updated.

File: sdk/eventhub/azure-eventhubs/README.md (hunk 11 above)
Comment (yunhaoling) [reply sdk/eventhub/azure-eventhubs/README.md]: thanks for the suggestion, doc is updated.

File: sdk/eventhub/azure-eventhubs/azure/eventhub/aio/client_async.py (hunk 5 above)
Comment (YijunXieMS) [reply sdk/eventhub/azure-eventhubs/azure/eventhub/aio/client_async.py]: Changed

File: sdk/eventhub/azure-eventhubs/azure/eventhub/client.py (hunk 6 above)
Comment (YijunXieMS) [reply sdk/eventhub/azure-eventhubs/azure/eventhub/client.py]: changed

File: sdk/eventhub/azure-eventhubs/azure/eventhub/client_abstract.py (hunk 7 above)
Comment (YijunXieMS) [reply sdk/eventhub/azure-eventhubs/azure/eventhub/client_abstract.py]: changed

**Pull request merged**